sdktests/server_side_persistence_consul.go

```diff
@@ -0,0 +1,121 @@
+package sdktests
+
+import (
+	"fmt"
+	"strings"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/launchdarkly/sdk-test-harness/v2/servicedef"
+)
+
+type ConsulPersistentStore struct {
+	consul *consul.Client
+}
+
+// {{{ PersistentStore implementation
+
+func (c ConsulPersistentStore) DSN() string {
+	// TODO: Fix this address lookup
+	return fmt.Sprintf("%s", consul.DefaultConfig().Address)
+}
+
+func (c *ConsulPersistentStore) Type() servicedef.SDKConfigPersistentType {
+	return servicedef.Consul
+}
+
+func (c *ConsulPersistentStore) Reset() error {
+	_, err := c.consul.KV().DeleteTree("/", nil)
+	return err
+}
+
+func (c *ConsulPersistentStore) Get(prefix, key string) (string, bool, error) {
+	kv := c.consul.KV()
+
+	pair, _, err := kv.Get(prefix+"/"+key, nil)
+	if err != nil || pair == nil {
+		return "", false, err
+	}
+
+	return string(pair.Value), true, nil
+}
+
+func (c *ConsulPersistentStore) GetMap(prefix, key string) (map[string]string, error) {
+	kv := c.consul.KV()
+	pairs, _, err := kv.List(prefix+"/"+key, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("list failed for %s: %s", key, err)
+	}
+
+	results := make(map[string]string)
+	for _, pair := range pairs {
+		flagKey := strings.TrimPrefix(pair.Key, prefix+"/"+key+"/")
+		results[flagKey] = string(pair.Value)
+	}
+	return results, nil
+}
+
+func (c *ConsulPersistentStore) WriteMap(prefix, key string, data map[string]string) error {
+	kv := c.consul.KV()
+
+	// Start by reading the existing keys; we will later delete any of these that weren't in allData.
+	pairs, _, err := kv.List(prefix, nil)
+	if err != nil {
+		return fmt.Errorf("failed to get existing items prior to Init: %s", err)
+	}
+	oldKeys := make(map[string]bool)
+	for _, p := range pairs {
+		oldKeys[p.Key] = true
+	}
+
+	ops := make([]*consul.KVTxnOp, 0)
+
+	for k, flag := range data {
+		op := &consul.KVTxnOp{Verb: consul.KVSet, Key: prefix + "/" + key + "/" + k, Value: []byte(flag)}
+		ops = append(ops, op)
+
+		oldKeys[prefix+"/"+key+"/"+k] = false
+	}
+
+	for k, v := range oldKeys {
+		if v {
+			op := &consul.KVTxnOp{Verb: consul.KVDelete, Key: prefix + "/" + key + "/" + k}
+			ops = append(ops, op)
+		}
+	}
+
+	// Submit all the queued operations, using as many transactions as needed. (We're not really using
+	// transactions for atomicity, since we're not atomic anyway if there's more than one transaction,
+	// but batching them reduces the number of calls to the server.)
+	err = batchOperations(kv, ops)
+	return err
+}
+
+func batchOperations(kv *consul.KV, ops []*consul.KVTxnOp) error {
+	for i := 0; i < len(ops); {
```

Comment:
I know this comes from the consul integration, but it would be good if we had a comment explaining what it is actually doing at a high level.